src/storage/src/render/upsert.rs

```diff
@@ -332,54 +285,54 @@ where
                 }
             }
         }
-        drop(previous_token);
-        while let Some(_event) = previous.next().await {
-            // Exchaust the previous input. It is expected to immediately reach the empty
-            // antichain since we have dropped its token.
-        }
-
-        consolidation::consolidate(&mut snapshot);
-
-        // The main key->value used to store previous values.
-        let mut state = StatsState::new(state().await, upsert_shared_metrics);
-
-        // A re-usable buffer of changes, per key. This is
-        // an `IndexMap` because it has to be `drain`-able
-        // and have a consistent iteration order.
-        let mut commands_state = indexmap::IndexMap::new();
-        let mut multi_get_scratch = Vec::new();
-
-        let mut initial_state_bytes = 0;
-        let mut initial_state_records = 0;
-
-        // Rehydrate the upsert state (and bump some stats), even if the snapshot is empty.
-        let snapshot = snapshot.into_iter().map(|((key, value), diff)| {
-            assert_eq!(diff, 1, "invalid upsert state");
-            initial_state_bytes = initial_state_bytes + ValueData::calculate_size(&value);
-            initial_state_records = initial_state_records + 1;
-
-            (key, Some(value))
-        });
-
-        let now = Instant::now();
-        let snapshot_size = snapshot.len();
-        state
-            .multi_put(snapshot)
+        // Flush out the rest of the snapshot.
+        stats += state
+            .merge_snapshot_chunk(update_buf.drain(..))
             .await
-            .expect("hashmap impl to not error");
+            .expect("hashmap impl to not fail");
+        drop(update_buf);
+        drop(batch_key_counter);
+
         upsert_metrics
             .rehydration_latency
             .set(now.elapsed().as_secs_f64());
+        upsert_metrics.rehydration_updates.set(stats.updates);
         upsert_metrics
             .rehydration_total
-            .set(u64::cast_from(snapshot_size));
-
+            .set(
+                stats
+                    .values_diff
+                    .try_into()
+                    .unwrap_or_else(|e: std::num::TryFromIntError| {
+                        tracing::warn!(
+                            "rehydration_total metric overflowed and is innacurate: {}",
```

Comment:
Here we convert from `i64` to `u64` which only fails when `values_diff` is negative and not due to overflow. When does this happen? Should we just return 0 or panic because there is a bug?

Reply:
They happen only if their are bugs. Bugs in counting should not bring the impl down, which is why we warn (which goes to sentry) and default to 0 here